sdk/samples/iot/hub/src/pnp_helper.c

```diff
@@ -0,0 +1,438 @@
+    {
+      return AZ_ERROR_ITEM_NOT_FOUND;
+    }
+  }
+
+  return AZ_ERROR_ITEM_NOT_FOUND;
+}
+
+// Check if the component name is in the model
+static az_result is_component_in_model(
+    az_span component_name,
+    const az_span** sample_components_ptr,
+    int32_t sample_components_num,
+    int32_t* out_index)
+{
+  int32_t index = 0;
+
+  if (az_span_ptr(component_name) == NULL || az_span_size(component_name) == 0)
+  {
+    return AZ_ERROR_UNEXPECTED_CHAR;
+  }
+
+  while (index < sample_components_num)
+  {
+    if ((az_span_size(component_name) == az_span_size(*sample_components_ptr[index]))
+        && (memcmp(
+                (void*)az_span_ptr(component_name),
+                (void*)az_span_ptr(*sample_components_ptr[index]),
+                (size_t)az_span_size(component_name))
+            == 0))
+    {
+      *out_index = index;
+      return AZ_OK;
+    }
+
+    index++;
+  }
+
+  return AZ_ERROR_UNEXPECTED_CHAR;
+}
+
+// Get the telemetry topic for PnP
+az_result pnp_helper_get_telemetry_topic(
+    az_iot_hub_client const* client,
+    az_iot_hub_client_properties* properties,
+    az_span component_name,
+    char* mqtt_topic,
+    size_t mqtt_topic_size,
+    size_t* out_mqtt_topic_length)
+{
+  az_result result;
+
+  az_iot_hub_client_properties pnp_properties;
+
+  if (properties == NULL)
+  {
+    AZ_RETURN_IF_FAILED(az_iot_hub_client_properties_init(
+        &pnp_properties, AZ_SPAN_FROM_BUFFER(pnp_properties_buffer), 0));
+  }
+
+  if ((result = az_iot_hub_client_properties_append(
+           properties != NULL ? properties : &pnp_properties,
+           component_telemetry_prop_span,
+           component_name))
+      != AZ_OK)
+  {
+    return result;
+  }
+  else
+  {
+    result = az_iot_hub_client_telemetry_get_publish_topic(
+        client,
+        properties != NULL ? properties : &pnp_properties,
+        mqtt_topic,
+        mqtt_topic_size,
+        out_mqtt_topic_length);
+  }
+
+  return result;
+}
+
+// Parse the component name and command name from a span
+az_result pnp_helper_parse_command_name(
+    az_span component_command,
+    az_span* component_name,
+    az_span* pnp_command_name)
+{
+  int32_t index = az_span_find(component_command, command_separator);
+  if (index > 0)
+  {
+    *component_name = az_span_slice(component_command, 0, index);
+    *pnp_command_name
+        = az_span_slice(component_command, index + 1, az_span_size(component_command));
+  }
+  else
+  {
+    *component_name = AZ_SPAN_NULL;
+    *pnp_command_name = component_command;
+  }
+  return AZ_OK;
+}
+
+// Create a reported property payload
+az_result pnp_helper_create_reported_property(
+    az_span json_buffer,
+    az_span component_name,
+    az_span property_name,
+    pnp_append_property_callback append_callback,
+    void* context,
+    az_span* out_span)
+{
+  az_result result;
+
+  az_json_writer json_writer;
+  result = az_json_writer_init(&json_writer, json_buffer, NULL);
```

Comment:
>result [](start = 2, length = 6)

what if it fails?

Reply:
good point wrapped in `AZ_RETURN_IF_FAILED`